### PR TITLE
roachtest: separate-process deployments in mixedversion

### PR DESCRIPTION
**Pull request by renatolabs**
This commit adds support `separate-process` deployments in
`mixedversion` tests. It is a natural enhancement of previous work
where we introduced `shared-process` deployments.

When a test runs in a `separate-process` deployment mode, each upgrade
will involve restarting the binaries for both the system tenant (also
known as the storage cluster), as well as for the
application (non-system) tenant. The storage cluster needs to finalize
its upgrade before the application tenant starts upgrading.

Mixed-version functions can be called while the system tenant is
upgrading and/or while the application tenant is upgrading.

For simplicity, only one separate-process tenant is created, and
handles traffic generated by the test. In addition, a SQL server is
started on every node where we run a storage cluster binary. In other
words, the two processes are colocated (without any form of resource
allocation) on the same nodes.

For now, separate-process deployments are disabled on a number of
tests where it's known to fail. We should, over time, investigate
these failures and re-enable the new deployment mode on them.

Epic: none

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
Marking this as ready for review. As with the shared-process work, it's a rather large PR that is difficult to break down in smaller pieces. Happy to do a walkthrough if there's interest.

The first 3 commits are small follow up fixes for recent PRs. They weren't caught because it was code in preparation for this `separate-process` work but not really exercised on master. The bulk of this PR is the last commit.

`separate-process` is currently only enabled in a few tests. It's disabled in tests that run locally (CI) for now, to avoid disruptions. In those cases, we only enable it in nightlies, where the issue can be analyzed using the standard TC artifacts.

There's a (currently in-progress) [TC run](https://teamcity.cockroachdb.com/viewLog.html?buildId=16782573&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel&tab=buildResultsDiv&branch_Cockroach_Nightlies=130194) on this branch, running every mixed-version test 5 times. The failures so far are "expected" -- the `acceptance/version-upgrade` failure specifically seems to happen with some regularity on the backup step. I'm planning to have DR take a look at it once this branch is merged.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 1]
```diff
@@ -192,11 +192,20 @@ const (
 	LastUpgradeStage
 	RunningUpgradeMigrationsStage
 	AfterUpgradeFinalizedStage
+	UpgradingSystemStage
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:195]: Do these two upgrading stages still adhere to `Upgrade stages are defined in the order they happen during test runs`? 

From my understanding, if a tenant is on `UpgradingSystemStage`, then it hasn't yet started any upgrades so this stage should come earlier. But I could also see why making comparisons against these two stages don't make sense.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 2]
```diff
@@ -254,31 +263,81 @@ var planMutators = []mutator{
...
+	) []testStep {
+		if p.deploymentMode == SeparateProcessDeployment {
+			// Separate-process deployments are different because the two
+			// services involved (system and tenant) restart at different
+			// times (since they run separate binaries): the system (storage
+			// cluster) needs to upgrade to completion first, and only then
+			// the tenant is allowed to start upgrading.
+			//
+			// To allow for internal and external hooks to recognize this
+			// stage, we set update the service's stage accordingly below.
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:277]: nit: I understand what's going on now, but it took me a couple rereads to get why we set the tenant to `UpgradingSystem` and system to `UpgradingTenant`. Maybe a mention that this stage means we're waiting for the system/tenant to finish upgrading might've nudged me in the right direction faster.

**Review by DarrylWong — COMMENTED** [review]
Not fully done reviewing, but publishing initial comments since I'll be OOO

File: pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go [hunk 3]
```diff
@@ -487,41 +432,108 @@ func (tr *testRunner) logStep(prefix string, step *singleStep, l *logger.Logger)
...
+
+	if tenantRuntime != nil {
+		if deploymentMode == SeparateProcessDeployment {
+			tw.AddRow(
+				withLabel("released versions", tenantRuntime.descriptor.Name),
+				toString(tenantReleasedVersions)...,
+			)
+
+			tw.AddRow(
+				withLabel("logical versions", tenantRuntime.descriptor.Name),
```
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go]: Good point. I used to call these "logical binary versions", as they are "binary" (i.e., related to the executable) but also logical (i.e., not the released `v24.x.x`). But it seems that in our documentation we refer to these as just binary versions. They are both (system and tenant) labeled as such now.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go [hunk 4]
```diff
@@ -487,41 +432,108 @@ func (tr *testRunner) logStep(prefix string, step *singleStep, l *logger.Logger)
...
+
+	// Released and logical binary versions are only per-tenant if we
+	// are in an separate-process deployment: otherwise, they are
+	// shared properties of system and tenants.
+	tw.AddRow(
+		withSystemLabel("released versions", deploymentMode == SeparateProcessDeployment),
+		toString(systemReleasedVersions)...,
+	)
+	tw.AddRow(
+		withSystemLabel("logical binary versions", deploymentMode == SeparateProcessDeployment),
```
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go]: Same as https://github.com/cockroachdb/cockroach/pull/130194#discussion_r1759028645.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 1 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go]: Good catch. As you said, it doesn't really make sense to perform any kind of comparison with these stages. I gave them an arbitrary large value (so that `<` comparisons are consistently false) and updated the comments to reflect this.

For that matter, I only expect mutators to need to do things that are low level enough to involve comparing stages in this way.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 2 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go]: I updated the comments slightly here, hopefully it made it a little clearer.

File: pkg/cmd/roachtest/tests/mixed_version_backup.go [hunk 5]
```diff
@@ -2647,6 +2664,12 @@ func registerBackupMixedVersion(r registry.Registry) {
 				// attempted.
 				mixedversion.UpgradeTimeout(30*time.Minute),
 				mixedversion.AlwaysUseLatestPredecessors,
+				// This test sometimes flake on separate-process
```
Comment (herkolategan) [inline pkg/cmd/roachtest/tests/mixed_version_backup.go:2667]: Nit: I'm wondering if we should have a meta issue / or issues that lists these tests, that are incompatible for unknown reasons, so we don't lose track?

**Review by herkolategan — APPROVED** [review]
**Great work!**

File: pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go [hunk 6]
```diff
@@ -386,55 +364,22 @@ func (tr *testRunner) stepError(
...
-
-		return loadAtomicVersions(service.clusterVersions)
-	}
-
-	systemDetails.clusterVersionsAfter = currentClusterVersions(tr.systemService)
-	if tenantDetails != nil {
-		tenantDetails.clusterVersionsAfter = currentClusterVersions(tr.tenantService)
+	lines := []string{
+		"test failed:",
+		fmt.Sprintf("test random seed: %d\n", tr.plan.seed),
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go:369]: Not related to this PR, but should we explicitly mention that the seed can be set through `COCKROACH_RANDOM_SEED`? Since we saw someone ask about the diff between this and `global-seed`

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go [hunk 7]
```diff
@@ -339,6 +339,17 @@ func newRand() *rand.Rand {
 
 func newTest(options ...CustomOption) *Test {
 	testOptions := defaultTestOptions()
+	defaultTestOverrides := []CustomOption{
+		// Enforce system-only deployments by default in tests; those that
+		// test multitenant deployments specifically should pass the
+		// `EnabledDeploymentModes` option explicitly.
+		EnabledDeploymentModes(SystemOnlyDeployment),
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:346]: Should we add `DisableSkipVersionUpgrades` to this and get rid of [the override here](https://github.com/cockroachdb/cockroach/blob/0f2ffbf8ccd3376dc04274a3d738a79626e4a26c/pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go#L89)?

**Review by DarrylWong — APPROVED** [review]
Left a few more small comments but nothing blocking. Besides that this LGTM!

File: pkg/cmd/roachtest/tests/versionupgrade.go [hunk 8]
```diff
@@ -108,7 +108,18 @@ func runVersionUpgrade(ctx context.Context, t test.Test, c cluster.Cluster) {
 		var cancel context.CancelFunc
 		testCtx, cancel = context.WithTimeout(ctx, localTimeout)
 		defer cancel()
-		opts = append(opts, mixedversion.NumUpgrades(1))
+		opts = append(
+			opts,
+			mixedversion.NumUpgrades(1),
+			// Disable separate-proces deployments in local runs, as it is
```
Comment (srosenberg) [inline pkg/cmd/roachtest/tests/versionupgrade.go:114]: Do we have an existing GH issue to reference here?

File: pkg/cmd/roachtest/tests/validate_system_schema_after_version_upgrade.go [hunk 9]
```diff
@@ -118,14 +118,28 @@ func runValidateSystemSchemaAfterVersionUpgrade(
...
 		mixedversion.NeverUseFixtures,
-	)
+	}
+
+	if c.IsLocal() {
+		opts = append(opts,
+			// Separate-process deployments are still in its early
+			// days. Disable it in local clusters (such as in CI) to avoid
+			// disruptions.
+			mixedversion.EnabledDeploymentModes(
```
Comment (srosenberg) [inline pkg/cmd/roachtest/tests/validate_system_schema_after_version_upgrade.go:136]: Maybe drop a `TODO` so it's easier to find.

File: pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go [hunk 10]
```diff
@@ -368,19 +368,26 @@ func RestartNodesWithNewBinary(
 	rand.Shuffle(len(nodes), func(i, j int) {
 		nodes[i], nodes[j] = nodes[j], nodes[i]
 	})
+
+	// Stop the cockroach process gracefully in order to drain it properly.
+	// This makes the upgrade closer to how users do it in production, but
+	// it's also needed to eliminate flakiness. In particular, this will
+	// make sure that DistSQL draining information is dissipated through
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go:375]: Nit: maybe `communicated`? `dissipated` is synonymous with `evaporated` :)

File: pkg/cmd/roachtest/option/options.go [hunk 11]
```diff
@@ -174,6 +174,17 @@ func SkipInit(opts interface{}) {
 	}
 }
 
+// Tag sets a process tag when stopping processes. Useful if we want
+// to kill a cockroach process that was started with `settings.TagOption`.
```
Comment (srosenberg) [inline pkg/cmd/roachtest/option/options.go:178]: Is `ProcessTag` actually used by the kill script? I couldn't find any reference sites other than `getClusterFromCache`.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go [hunk 12]
```diff
@@ -180,21 +185,21 @@ var (
...
-	// component of tests generated by the framework.
+	// OldestSupportedVersionSP is similar to `OldestSupportedVersion`,
+	// but applies only to shared-process virtual cluster deployments.
+	// The reason it is different for now is that v23.1.5 is the first
+	// version to ship with the ability to change the default target
+	// cluster, a core component of tests generated by the framework.
 	//
 	// TODO(testeng): remove this constant when 23.1 reaches EOL.
-	OldestSupportedVersionVC = clusterupgrade.MustParseVersion("v23.1.5")
+	OldestSupportedVersionSP = clusterupgrade.MustParseVersion("v23.1.5")
```
[the remark this answers is not in the document]
Comment (srosenberg) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:195]: To avoid (name) overloading, we could also consider using `external process`. After all, our unit tests still use that terminology, e.g.,

> test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 13]
```diff
@@ -254,31 +269,85 @@ var planMutators = []mutator{
...
+		for _, s := range p.services() {
+			s.startUpgrade(upgrade.to)
+		}
+
+		systemUpgradeSteps := upgradeStepsForService(
+			p.currentContext.System, upgrade.from, upgrade.to, virtualClusterSetup, scheduleHooks,
+		)
+
+		if p.deploymentMode != SeparateProcessDeployment || !virtualClusterSetup {
+			upgrade.Add(systemUpgradeSteps)
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:335]: I had to pause here to remember that `systemUpgradeSteps` is essentially equivalent to non-multitenant upgrades; it might be worth dropping an explicit comment as a reminder.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 14]
```diff
@@ -446,6 +520,12 @@ func (p *testPlanner) setupTest() (testSetup, []*upgradePlan) {
 
 	var tenantSetup *serviceSetup
 	if p.isMultitenant() {
+		for _, n := range p.currentContext.Tenant.Descriptor.Nodes {
+			handleInternalError(
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:524]: Remind me why `changeVersion` here wasn't necessary before this PR? Previously, `changeVersion` was confined to `changeVersionSteps`.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go [hunk 15]
```diff
@@ -1081,15 +1095,12 @@ func (h hooks) Filter(testContext Context) hooks {
...
-// one stop channel (`shouldStop`) for each hook.
-func (h hooks) AsSteps(
-	label string, prng *rand.Rand, testContext *Context, stopChans []shouldStop, isLocal bool,
-) []testStep {
+// AsSteps transforms the sequence of hooks into a sequence of
+// `*singleStep` structs to be run in some way by the planner
+// (sequentially, concurrently, etc). `stopChans` should either be
+// `nil` for steps that are not meant to be run in the background, or
+// contain one stop channel (`shouldStop`) for each hook.
+func (h hooks) AsSteps(prng *rand.Rand, testContext *Context, stopChans []shouldStop) []testStep {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:1103]: Nice! I like that `concurrently` concern was moved to the caller(s).

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 16]
```diff
@@ -681,90 +833,121 @@ func (p *testPlanner) changeVersionSteps(
...
 	))
 
-	// At this point, we just upgraded the storage cluster; we might
-	// need to patch up the `system.tenant_settings` table in order for
-	// the tenant upgrade to work.
+	// If this is a system-only or separate-process deployment, we are
+	// done -- the `service` passed is upgraded at this point. However,
+	// for shared-process deployments, we also need to upgrade the
+	// underlying tenant.
+	if p.deploymentMode == SystemOnlyDeployment || p.deploymentMode == SeparateProcessDeployment {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:898]: Nit: I find it confusing to have to switch between `p.deploymentMode`, `p.isMultitenant`, and `virtualClusterSetup`, sometimes in the same func. Perhaps adding to the confusion is the fact that `p.isMultitenant` and `virtualClusterSetup` are _not_ equivalent under all invocations. It's not immediately clear if there is a way to simplify; perhaps a top-level comment summarizing how they're used wrt (FSM) state transitions.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 17]
```diff
@@ -177,11 +177,12 @@ const (
 
 const (
 	// Upgrade stages are defined in the order they happen during test
-	// runs, so that we are able to select, for example, "stages after
-	// rollback" by doing `stage > RollbackUpgrade`.  Note that
-	// `BackgroundStage` is special in the sense that it may continue to
-	// run across stage changes, so doing direct stage comparisons as
-	// mentioned above doesn't make sense for background functions.
+	// runs, so that we are able to implementat mutators that select,
```
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go]: Fixed.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:195]: Given that this variable is only used in a couple places internally and it will go away when 23.1 reaches EOL (in a month or two), I'm leaning towards keeping this as-is until then.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go (hunk 7 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:338]: Good point, done.

File: pkg/cmd/roachtest/tests/versionupgrade.go (hunk 8 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/tests/versionupgrade.go:114]: We don't. I expect this test to fail in nightly runs shortly after it's merged. When it does, I want to ask DR to help debug the issue, given we'll have all the artifacts at that point.

Depending on how that goes we can then create a new issue, or fix something in DR land, or just skip separate-process entirely for this test.

File: pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go (hunk 10 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go]: Updated. I just moved this comment in the PR, it has been around for a while I believe.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 13 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:338]: Comment added.

File: pkg/cmd/roachtest/tests/validate_system_schema_after_version_upgrade.go (hunk 9 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/tests/validate_system_schema_after_version_upgrade.go:139]: Done.

File: pkg/cmd/roachtest/option/options.go (hunk 11 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/option/options.go:193]: It's fairly subtle, but this is how it works:

* `c.Stop` is called with the `Tag` option, causing `RoachprodOpts.ProcessTag` to be set (the code you're commenting on)
* Those `RoachprodOpts` are passed to `roachprod.Stop`:

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/cmd/roachtest/cluster.go#L2397

* The process tag is passed when loading the cluster from the cache:

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/roachprod.go#L912

* This means that the `*install.SyncedCluster` instance will create a `ClusterSettings` object with that tag, and `SyncedCluster` embeds `ClusterSettings`:

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_settings.go#L112-L126

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_synced.go#L74-L80

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_synced.go#L62

* When the `Stop` call eventually calls `c.kill()`, the kill script will filter the process to kill based on the `roachprodEnvRegex`:

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_synced.go#L532-L535

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_synced.go#L543

* `roachprodEnvRegex` returns the value that should be set in the `ROACHPROD` environment variable on the process. It includes the label we set:

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_synced.go#L292-L294

https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/roachprod/install/cluster_synced.go#L270-L280

*****
TLDR: only a process started with the same tag should be killed when `c.Stop()` is called with `option.Tag()`.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 14 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:534]: This was already "broken" before, but it was inconsequential before the introduction of separate-process tenants.

What this is doing is updating the context associated with the tenant to set the actual released version it is bootstrapped in. The context is initialized in [here](https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/cmd/roachtest/roachtestutil/mixedversion/context.go#L168-L199), and the initial version is set to the first binary used in the upgrade test (the one used to [bootstrap the storage cluster](https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go#L757)). However, tenants are (or can be) started later, potentially after the storage cluster has already gone through a few upgrades.

In separate-process deployments, we display the "released version" for each node in the cluster. Without this change, the version displayed would have been potentially incorrect if bootstrap version != initial version for storage cluster.

Hopefully that helps understand. I added a comment to this code with similar information.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go (hunk 15 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:1113]: Yeah, having this function return a concurrent step was a mistake, it's the planner's responsibility to decide how to run steps.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 16 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:910]: Noted.

If it helps:

* `p.deploymentMode` is the actual deployment mode we're generating a plan for. Typically used when we want to do something specific to a single deployment mode (i.e., `SharedProcess` in this case).
* `p.isMultitenant()` is just a helper to group common logic shared by shared- and separate-process multi-tenant deployments.
* `virtualClusterSetup`: admittedly, this is a poorly named variable. It encodes whether the tenant in the test (shared or separate) is "setup" -- i.e., if it's already expected to be running at that point in the plan. Maybe `virtualClusterRunning` is a better name.

File: pkg/cmd/roachtest/tests/mixed_version_backup.go (hunk 5 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/tests/mixed_version_backup.go:2667]: Created #130968.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go (hunk 6 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go]: Done.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 16 above)
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:910]: Renamed `virtualClusterSetup` => `virtualClusterRunning`.

**Comment by renatolabs** [comment]
[Another run](https://teamcity.cockroachdb.com/viewLog.html?buildId=16946036&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel&tab=buildResultsDiv) on this branch reveals that things are still working.

The `rebalance` and `follower-reads` failures happen on master too (on shared-process deployments). The `tpcc` and `acceptance` failures are known and to be debugged once this PR merges.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 18]
```diff
@@ -683,90 +845,121 @@ func (p *testPlanner) changeVersionSteps(
 // hooks while the cluster version is changing. At the end of this
 // process, we wait for all the migrations to finish running.
 func (p *testPlanner) finalizeUpgradeSteps(
-	fromVersion, toVersion *clusterupgrade.Version, scheduleHooks, virtualClusterSetup bool,
+	service *ServiceContext,
+	fromVersion, toVersion *clusterupgrade.Version,
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:849]: `fromVersion` is no longer used

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 19]
```diff
@@ -683,90 +845,121 @@ func (p *testPlanner) changeVersionSteps(
...
 			desiredVersion:     versionToClusterVersion(toVersion),
-			virtualClusterName: install.SystemInterfaceName,
+			virtualClusterName: service.Descriptor.Name,
 		},
 	))
 
-	// At this point, we just upgraded the storage cluster; we might
-	// need to patch up the `system.tenant_settings` table in order for
-	// the tenant upgrade to work.
+	// If this is a system-only or separate-process deployment, we are
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:906]: Nit: I found the "we are done" part confusing in case of separate-process. The fact is that this function is invoked twice, i.e., first with `p.currentContext.System`, then with `p.currentContext.Tenant`. It might be worth clarifying.

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 20]
```diff
@@ -683,90 +845,121 @@ func (p *testPlanner) changeVersionSteps(
...
+	// done -- the `service` passed is upgraded at this point. However,
+	// for shared-process deployments, we also need to upgrade the
+	// underlying tenant.
+	if p.deploymentMode == SystemOnlyDeployment || p.deploymentMode == SeparateProcessDeployment {
+		return steps
+	}
+
+	// At this point, we need to upgrade the shared-process tenant. We
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:914]: Nit: weird asymmetry since it's one of the only (?) cases where we seemingly have to do more work in the case of shared-process. Of course, in reality, separate-process requires two invocations of `finalizeUpgradeSteps` whereas shared-process only one, hence the asymmetry.

**Review by srosenberg — APPROVED** [review]

File: pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go [hunk 21]
```diff
@@ -613,7 +632,10 @@ func (tr *testRunner) refreshServiceData(ctx context.Context, service *serviceRu
 		return err
 	}
 
-	tr.monitor.Init()
+	if tr.plan.deploymentMode == SeparateProcessDeployment && service.descriptor.Name != install.SystemInterfaceName {
+		tr.monitor.Init()
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go:636]: Who monitors the system tenant now in case of shared-process and single-tenant?

Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go:650]: Yikes, good catch! The intended logic is to only start the monitor once every cockroach process is running; in the case of separate-process, this is equivalent to checking if the _tenant_ is running (since system is always started first).

Fixed.

**Comment by renatolabs** [comment]
Ran a final [build](https://teamcity.cockroachdb.com/viewLog.html?buildId=17058873&) on this branch, the failures are either expected or unrelated.

TFTRs!

bors r=herkolategan,DarrylWong,srosenberg
